### PR TITLE
Add labels to generated OpenShift resources

### PR DESCRIPTION
**Pull request by jafarre-viewnext**
Closes https://github.com/opendevstack/ods-quickstarters/issues/512

### CODE REVIEW

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (michaelsauter) [inline src/org/ods/openshift/OpenShiftResourceMetadata.groovy:18]: Please add `@TypeChecked`, that is really helpful.

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (michaelsauter) [inline src/org/ods/openshift/OpenShiftResourceMetadata.groovy:57]: maybe name it `labelToMetadataMapping`?

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (michaelsauter) [inline src/org/ods/openshift/OpenShiftResourceMetadata.groovy:429]: Why do we read from chart here? why not let that be handled by Helm itself?

File: src/org/ods/services/OpenShiftService.groovy
Comment (michaelsauter) [inline src/org/ods/services/OpenShiftService.groovy:650]: This might trigger a redeploy, right?

**Review by michaelsauter — COMMENTED** [review]
Just glanced over it. Nice work.

I think in general the relationship between tools like Helm and this thing here is not entirely clear to me. Plus labelling causes a redeploy if I'm not mistaken which is not ideal I think?

**Comment by michaelsauter** [comment]
In general, I am a bit uneasy with the perceived complexity this introduces. Here's where I'm coming from:

Mid-term, I think we should migrate to Helm templating. Together with this, I think we should / need to cut the automatic export functionality (there is no sane way to do it IMHO). As a consequence everyone would need to define their K8S resources in version control. I strongly believe this is the right way to go.
However with this on the horizon, I wonder how that would interact with the feature proposed here. Maybe there is a way to integrate labelling in a different way, e.g. through Helm? Or maybe only check for labels, instead of applying them in a separate process?

Just thinking out loud here ...

**Comment by metmajer** [comment]
@michaelsauter I am not clear yet in which ways Helm would change the rules of the game compared to Quickstarters. Independent of this, components should always apply a set of labels we deem necessary, and the Release Manager would add some extra ones on top. As long as `app.kubernetes.io` labels are applied in a Helm-conform way, we should be future-proof, no?

**Comment by michaelsauter** [comment]
@metmajer I think what needs to happen is that quickstarters ship with a Helm chart. Meaning we would keep the Helm templates under version control right from the start. We could by default set all the labels as desired by the platform. Then later on only a check is needed?

File: src/org/ods/services/OpenShiftService.groovy
Comment (jafarre-viewnext) [reply src/org/ods/services/OpenShiftService.groovy:650]: In all tests I have done, labeling does not trigger redeploys. Otherwise, it is just a question of temporarily disabling automatic rollouts, but it doesn't seem to be needed at all.

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (jafarre-viewnext) [reply src/org/ods/openshift/OpenShiftResourceMetadata.groovy:429]: I considered the possibility to integrate with the helm tool and even get several label values directly it, but I left that as out of scope (actually, setting this helm-related chart is an extra).
This was so easy to do, that it felt more straightforward just to read the chart file as a YAML. It can be improved in the future in quite a tranparent way, I would say.
If you don't feel confortable with my approach, I can disable this label for the moment or see how long it would take to integrate with the helm tool.

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (jafarre-viewnext) [reply src/org/ods/openshift/OpenShiftResourceMetadata.groovy:57]: No problem

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (jafarre-viewnext) [reply src/org/ods/openshift/OpenShiftResourceMetadata.groovy:18]: Will do.

**Comment by jafarre-viewnext** [comment]
> In general, I am a bit uneasy with the perceived complexity this introduces. Here's where I'm coming from:
> 
> Mid-term, I think we should migrate to Helm templating. Together with this, I think we should / need to cut the automatic export functionality (there is no sane way to do it IMHO). As a consequence everyone would need to define their K8S resources in version control. I strongly believe this is the right way to go.
> However with this on the horizon, I wonder how that would interact with the feature proposed here. Maybe there is a way to integrate labelling in a different way, e.g. through Helm? Or maybe only check for labels, instead of applying them in a separate process?
> 
> Just thinking out loud here ...

Helm suggests how to set the recommended labels in the templates. It can be done that way in the future, or the shared library could use the helm tool to get the suggested template variable labels and apply them like it's being done in this PR.
Do you foresee any possible backwards-compatibility issues?

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (jafarre-viewnext) [reply src/org/ods/openshift/OpenShiftResourceMetadata.groovy:18]: Done

File: src/org/ods/openshift/OpenShiftResourceMetadata.groovy
Comment (jafarre-viewnext) [reply src/org/ods/openshift/OpenShiftResourceMetadata.groovy]: Done

**Comment by jafarre-viewnext** [comment]
@michaelsauter ,@metmajer , when you have time, please, I need approval for the latest changes. Thank you.

**Review by metmajer — APPROVED** [review]

**Pull request merged**
